src/arith/transitive_comparison_analyzer.cc

```diff
@@ -0,0 +1,683 @@
+                                 !expr_equal(range->extent, (*it).second->extent);
+    if (differs_from_previous) {
+      ICHECK(allow_override) << "Binding of variable " << var << " as " << range
+                             << " conflicts with previous binding as " << (*it).second;
+      if (auto key = ExprToPreviousKey(var)) {
+        knowns_.erase(std::remove_if(knowns_.begin(), knowns_.end(),
+                                     [&](const auto& known) { return known.lhs_ == key.value(); }),
+                      knowns_.end());
+      }
+    }
+  }
+
+  prev_bindings_.Set(var, range);
+
+  if (is_const_int(range->extent, 1)) {
+    AddKnown(var == range->min, &knowns_);
+  } else {
+    AddKnown(var >= range->min, &knowns_);
+    AddKnown(var < range->min + range->extent, &knowns_);
+  }
+}
+
+void TransitiveComparisonAnalyzer::Impl::Bind(const tir::Var& var, const PrimExpr& expr,
+                                              bool allow_override) {
+  Bind(var, Range::FromMinExtent(expr, 1), allow_override);
+}
+
+std::function<void()> TransitiveComparisonAnalyzer::Impl::EnterConstraint(const PrimExpr& expr) {
+  size_t old_literal_size = scoped_knowns_.size();
+  AddKnown(expr, &scoped_knowns_);
+  size_t new_literal_size = scoped_knowns_.size();
+
+  PrimExpr temp = expr;
+  auto frecover = [old_literal_size, new_literal_size, this, temp]() {
+    ICHECK_EQ(scoped_knowns_.size(), new_literal_size);
+    scoped_knowns_.erase(scoped_knowns_.begin() + old_literal_size, scoped_knowns_.end());
+  };
+  return frecover;
+}
+
+CompareResult TransitiveComparisonAnalyzer::Impl::TryCompare(const PrimExpr& lhs_expr,
+                                                             const PrimExpr& rhs_expr) const {
+  // Currently only supports integer checks
+  if (!lhs_expr.dtype().is_int() || !rhs_expr.dtype().is_int()) {
+    return CompareResult::kUnknown;
+  }
+
+  // Bail out early if possible.  This int check should have been
+  // constant-folded earlier, so this check shouldn't occur.
+  auto* x_int = lhs_expr.as<IntImmNode>();
+  auto* y_int = rhs_expr.as<IntImmNode>();
+  if (x_int && y_int) {
+    if (x_int->value < y_int->value) {
+      return CompareResult::kLT;
+    } else if (x_int->value > y_int->value) {
+      return CompareResult::kGT;
+    } else {
+      return CompareResult::kEQ;
+    }
+  }
+
+  auto [lhs, rhs, offset] = ExtractOffsets(lhs_expr, rhs_expr);
+  auto lhs_key = ExprToPreviousKey(lhs);
+  auto rhs_key = ExprToPreviousKey(rhs);
+
+  if (!lhs_key.has_value() || !rhs_key.has_value()) {
+    return CompareResult::kUnknown;
+  }
+
+  auto from_lhs = TryCompareFromLHS(lhs_key.value(), rhs_key.value(), offset, lhs, rhs);
+  auto from_rhs = Reverse(TryCompareFromLHS(rhs_key.value(), lhs_key.value(), -offset, rhs, lhs));
+  auto output = from_lhs & from_rhs;
+
+  return output;
+}
+
+CompareResult TransitiveComparisonAnalyzer::Impl::TryCompareFromLHS(
```

Comment:
nit: TryCompareFromLHS is a bit long

Reply:
Hmm, good point.  Renamed to `DFSFromLHS`, which hopefully works with the updated documentation to reduce the verbosity.

Reply:
Oh shoot sorry I was meaning the function length was quite long and could be easier to understand if factored but it was completely a nitpick and not necessary given everything else looks good now. But I do like the new name too!